### DISCUSSION TITLE
Testing OpenLDAP models

### DISCUSSION THREAD

**Discussion by gjm** [DirectoryTree/LdapRecord#321] [General]
Is there a way to test OpenLDAP models? Reading the docs I can only see some shallow docs to test authentication attempts.

I would like to test code that looks something like this:
```php
$user = new User(['dn' => 'john', 'cn' => 'John Doe']);
$user->save();

$user2 = User::find('dn=john,dc=local,dc=com');
$this->assertEquals($user, $user2);
```

Is there a way to do something like this?

**Comment by stevebauman** [comment]
Hi @gjm,

To perform identity checks against models, you can use the `is()` method:

```php
$user = new User(['dn' => 'john', 'cn' => 'John Doe']);
$user->save();

$user2 = User::find('dn=john,dc=local,dc=com');

$this->assertTrue($user->is($user2));
```

The `is()` method checks that the users distinguished name and connection name are identical (compared strictly).

Does this answer your question?

**Comment by gjm** [comment]
@stevebauman sorry I probably didn't make myself clear. I want to fake the connections. I want to be able to test it out without actually connecting to the server. I thought that was implied.

**Reply by stevebauman** [reply]
Hi @gjm, apologies for the late reply on this issue.

For a robust LDAP test environment, you'll have to use LdapRecord-Laravel's [Directory Emulator](https://ldaprecord.com/docs/laravel/v2/testing/#directory-emulator). It provides the easiest test suite, since you can utilize LdapRecord's API directly for creating test models.

If you're not using Laravel, you'll have to utilize LdapRecord's connection fakes. Connection fakes allow you to add expected LDAP method calls and return a dummy responses.

I haven't had the time to create extensive documentation for them, but you can browse the LdapRecord test suite and see their usage:

https://github.com/DirectoryTree/LdapRecord/blob/master/tests/Models/ModelQueryTest.php

```php
use LdapRecord\Testing\LdapFake;
use LdapRecord\Testing\DirectoryFake;
use LdapRecord\Models\ActiveDirectory\User;

$fake = DirectoryFake::setup();

$fake->getLdapConnection()->expect(
    LdapFake::operation('search')->andReturn('...')
);

$results = User::all();
```

Hope this helps!

**Reply by gjm** [reply]
Thanks @stevebauman. Although I'm using Laravel I'm using the "general purpose" LdapRecord so I'll look into the connection fakes.

Thanks again.
